File: src/main/java/com/yauhenl/foe/service/RequestService.java
package com.yauhenl.foe.service;

import org.bson.Document;
import org.json.JSONArray;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.xml.bind.DatatypeConverter;
import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;
import java.util.stream.Stream;

import static java.lang.String.format;

@Service
public class RequestService {
    private static final Logger logger = LoggerFactory.getLogger(RequestService.class);
    private static final Random rnd = new Random();

    @Value("${game.server}")
    private String server;

    @Value("${game.secret}")
    private String secret;

    @Value("${game.timestamp}")
    private String timestamp;

    @Value("${game.version}")
    private String version;

    private Integer requestId = rnd.nextInt(255);

    public Document findByRequestMethod(List<Document> data, String mathodName) {
        return data.stream().filter(it -> mathodName.equals(it.getString("requestMethod"))).findAny().map(it -> it.get("responseData", Document.class)).orElse(null);
    }

    public List<Document> findListByRequestMethod(List<Document> data, String mathodName) {
        return data.stream().filter(it -> mathodName.equals(it.getString("requestMethod"))).findAny().map(it -> it.get("responseData", List.class)).orElse(null);
    }

    public List<Document> postRequest(JSONArray payload, String sid, String userKey) {
        String url = format("https://%s.forgeofempires.com/game/json?h=%s", server, userKey);
        payload.getJSONObject(0).put("requestId", requestId++);
        String body = getBody(payload);
        Map<String, String> headers = new HashMap<>();
        headers.put("Accept", "*/*");
        headers.put("Accept-Encoding", "gzip, deflate, br");
        headers.put("Accept-Language", "en,ru;q=0.8");
        headers.put("Client-Identification", format("version=%s; requiredVersion=%s; platform=bro; platformVersion=web", version, version));
        headers.put("Connection", "keep-alive");
        headers.put("Content-Type", "application/json");
        headers.put("Cookie", format("portal_ref_session=1; metricsUvId=ffcf4c04-e10a-42a8-bc83-dadf16cc2c70; sid=%s; req_page_info=game_v1; start_page_type=game; start_page_version=v1; ref=daa_ru_cis; ig_conv_last_site=https://%s.forgeofempires.com/game/index", sid, server));
        headers.put("Host", format("%s.forgeofempires.com", server));
        headers.put("Origin", "https://foeru.innogamescdn.com");
        headers.put("Referer", format("https://foeru.innogamescdn.com/swf/Preloader.swf?%s/[[DYNAMIC]]/1", timestamp));
        headers.put("Signature", getSignature(body, userKey));
        headers.put("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36");
        headers.put("X-Requested-With", "ShockwaveFlash/27.0.0.183");
//        JSONArray result = post(url, headers, new HashMap<>(), body).getJsonArray();
        StringBuilder data = new StringBuilder();
        try {
            Path path = Paths.get(Objects.requireNonNull(getClass().getClassLoader().getResource("response.json")).toURI());
            Stream<String> lines = Files.lines(path);
            lines.forEach(line -> data.append(line).append("\n"));
            lines.close();
        } catch (URISyntaxException | IOException e) {
            logger.error(e.getMessage());
        }
        JSONArray result = new JSONArray(data.toString().trim());
        List<Document> documents = new ArrayList<>();
        for (int i = 0; i < result.length(); i++) {
            documents.add(Document.parse(result.getJSONObject(i).toString()));
        }
        return documents;
    }

    private String getSignature(String body, String userKey) {
        String data = userKey + secret + body;
        MessageDigest md = null;
        try {
            md = MessageDigest.getInstance("MD5");
        } catch (NoSuchAlgorithmException e) {
            logger.error("NoSuchAlgorithmException", e);
        }
        byte[] digest = Objects.requireNonNull(md).digest(data.getBytes());
        return DatatypeConverter.printHexBinary(digest).substring(0, 10).toLowerCase();
    }

    private String getBody(JSONArray payload) {
        return payload.toString().replace(" ", "");
    }
}

File: src/main/kotlin/com/yauhenl/foe/service/RequestService.kt
package com.yauhenl.foe.service

import com.yauhenl.foe.model.Account
import khttp.post
import org.json.JSONArray
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.security.MessageDigest
import javax.xml.bind.DatatypeConverter
import org.apache.tomcat.jni.User.username
import org.json.JSONObject


@Service
class RequestService(
        @Value("\${game.server}") val server: String,
        @Value("\${game.secret}") val secret: String,
        @Value("\${login.user_key}") val userKey: String,
        @Value("\${login.sid}") val sid: String,
        @Value("\${game.timestamp}") val timestamp: String,
        @Value("\${game.version}") val version: String) {

    private var requestId = 0

    fun initialize() {
        val account = Account()
        val data = postRequest(account.getData(requestId++))
        account.data = service(data, "StartupService")
        account.data.put("taverns", method(data, "getOtherTavernStates"))
        account.data.put("resources", method(data, "getPlayerResources"))
    }

    fun service(data: JSONArray, service: String): JSONObject {
        (0..(data.length() - 1))
                .map { data.getJSONObject(it) }
                .filter { it.getString("requestClass") == service }
                .first { return it.getJSONObject("responseData") }
        return JSONObject()
    }

    fun method(data: JSONArray, method: String): JSONArray {
        (0..(data.length() - 1))
                .map { data.getJSONObject(it) }
                .filter { it.getString("requestMethod") == method }
                .first { return it.getJSONArray("responseData") }
        return JSONArray()
    }

    fun postRequest(payload: JSONArray): JSONArray {
        val url = "https://${server}.forgeofempires.com/game/json?h=${userKey}"
        val body = getBody(payload)
        val headers = hashMapOf(
                "Accept" to "*/*",
                "Accept-Encoding" to "gzip, deflate, br",
                "Accept-Language" to "en,ru;q=0.8",
                "Client-Identification" to "version=${version}; requiredVersion=${version}; platform=bro; platformVersion=web",
                "Connection" to "keep-alive",
                "Content-Type" to "application/json",
                "Cookie" to "portal_ref_session=1; metricsUvId=ffcf4c04-e10a-42a8-bc83-dadf16cc2c70; sid=${sid}; req_page_info=game_v1; start_page_type=game; start_page_version=v1; ref=daa_ru_cis; ig_conv_last_site=https://${server}.forgeofempires.com/game/index",
                "Host" to "${server}.forgeofempires.com",
                "Origin" to "https://foeru.innogamescdn.com",
                "Referer" to "https://foeru.innogamescdn.com/swf/Preloader.swf?${timestamp}/[[DYNAMIC]]/1",
                "Signature" to getSignature(body),
                "User-Agent" to "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36",
                "X-Requested-With" to "ShockwaveFlash/27.0.0.183")
        return post(url, headers, data = body).jsonArray
    }

    fun getSignature(body: String): String {
        val data = userKey + secret + body
        val md = MessageDigest.getInstance("MD5")
        val digest = md.digest(data.toByteArray())
        return DatatypeConverter.printHexBinary(digest).take(10).toLowerCase()
    }

    fun getBody(payload: JSONArray): String {
        return payload.toString().replace(" ", "")
    }
}
